x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1145,49 +1175,47 @@ private static LogicalPlan skipPlan(UnaryPlan plan) {
         return new LocalRelation(plan.source(), new EmptyExecutable(plan.output()));
     }
 
-    static class SkipQueryIfFoldingProjection extends OptimizerRule<LogicalPlan> {
-        @Override
-        protected LogicalPlan rule(LogicalPlan plan) {
-            List<LogicalPlan> leaves = plan.collectLeaves();
+    static class PushProjectionsIntoLocalRelation extends OptimizerRule<UnaryPlan> {
```

Comment:
I split the `SkipQueryIfFoldingProjection` transformation into `PushProjectionsIntoLocalRelation` and `SkipQueryForOnlyLiteralAggregations` because it looks like two different concerns and the combined logic becomes a bit more complicated with this PR.